### PR TITLE
Transform

### PR DESCRIPTION
**Pull request by BenGriffith**
1. Updated .env-template with defaults
2. Updated Makefile
3. Removed use of decorator along with tests
4. Implemented namedtuples instead of standard dictionaries
5. Added date_dim table load and name_dim table load

### CODE REVIEW

File: src/crypto/transform.py [hunk 1]
```diff
@@ -1,69 +1,135 @@
...
+        row = DateDim(
+            date_key=f"{d:%Y-%m-%d}",
+            year=d.year,
+            month_key=d.month,
+            day=d.day,
+            day_key=d.isoweekday(),
+            week_number=d.isocalendar().week,
+            week_end=d.fromisocalendar(d.year, d.isocalendar().week, 7).strftime('%Y-%m-%d'),
+            month_end=(d + relativedelta(day=31)).strftime("%Y-%m-%d")
+        )._asdict()
```
Comment (bbelderbos) [inline src/crypto/transform.py:91]: funny how the Standard Library is not always consistent: this method should probably be "public" :)

File: src/crypto/transform.py [hunk 2]
```diff
@@ -1,69 +1,135 @@
...
+        return [row]
     
-    def load_table(self, table_id: str, rows: list) -> None:
-        table = self.bq_client.get_table(table=table_id)
-        self.bq_client.insert_rows(table=table_id, rows=rows, selected_fields=table.schema)
-        
+    def _get_name_dim_keys(self):
+        query = """SELECT DISTINCT symbol FROM {name_dim}""".format(name_dim=NAME_DIM)
+        job = self.bq_client.query(query=query)
```
Comment (bbelderbos) [inline src/crypto/transform.py:96]: Seems a pass through (façade?), any reason this logic is here instead on bq_client? Just wondering ...

File: src/crypto/transform.py [hunk 3]
```diff
@@ -1,69 +1,135 @@
...
+        result = job.result()
+        return result
+
+    def name_dim_rows(self, crypto_data: dict) -> None:
+        result = self._get_name_dim_keys()
+        symbols = [row["symbol"] for row in result]
```
Comment (bbelderbos) [inline src/crypto/transform.py:102]: if you look it for lookup (skip below), I think a set would be more appropriate

File: src/crypto/utils/api.py [hunk 4]
```diff
@@ -16,6 +16,6 @@ def __init__(self, limit: int = 5000) -> None:
         }
 
     def request(self, url: str) -> dict:
-        response = requests.get(url, headers=self._headers, params=self._parameters)
+        response = requests.get(url, headers=self.headers, params=self.parameters)
```
Comment (bbelderbos) [inline src/crypto/utils/api.py:19]: I was wondering if I would add a raise for status, but last time I did not:
https://github.com/PyBites-Open-Source/search/blob/main/src/pybites_search/base.py#L40

File: src/crypto/utils/constants.py [hunk 5]
```diff
@@ -15,4 +15,6 @@
 DATASET = config("DATASET", cast=str)
 DAY_DIM = f"{PROJECT_ID}.{DATASET}.{config('DAY_DIM', cast=str)}"
 MONTH_DIM = f"{PROJECT_ID}.{DATASET}.{config('MONTH_DIM', cast=str)}"
+DATE_DIM = f"{PROJECT_ID}.{DATASET}.{config('DATE_DIM', cast=str)}"
```
Comment (bbelderbos) [inline src/crypto/utils/constants.py:18]: is that python-decouple? then I think str is already the default, no?

File: src/crypto/utils/setup.py [hunk 6]
```diff
@@ -0,0 +1,6 @@
+from collections import namedtuple
+
+DayDim = namedtuple(typename="DayDim", field_names="day_key name")
```
Comment (bbelderbos) [inline src/crypto/utils/setup.py:3]: Lately I always defined those with typing, easy to do FYI: https://www.youtube.com/watch?v=ljQHjQx76Po

**Review by bbelderbos — COMMENTED** [review]
Great work, talk later ...

File: src/crypto/transform.py [hunk 7]
```diff
@@ -1,69 +1,135 @@
...
     BUCKET,
     CLOUD_STORAGE,
     BIGQUERY,
+    PROJECT_ID,
     DAY_DIM,
     MONTH_DIM,
-    INITIAL_LOAD
+    DATE_DIM,
+    INITIAL_LOAD,
+    NAME_DIM,
```
[the remark this answers is not in the document]
Comment (BenGriffith) [reply src/crypto/transform.py:22]: yes, there is. That's a good suggestion. Feel like it'll clean it up.

File: src/crypto/transform.py [hunk 8]
```diff
@@ -1,69 +1,135 @@
...
-        return self.storage_client.bucket(bucket_name=self.bucket_name)
+        return self._bucket
     
     def blob(self, blob_name: str) -> Blob:
-        return self.bucket().blob(blob_name=blob_name)
+        return self.bucket.blob(blob_name=blob_name)
 
     def read_blob(self) -> dict:
-        blob_name = f"{datetime.today().strftime('%Y-%m-%d')}.json"
+        blob_name = f"{datetime.today().strftime('%Y-%m-%d')}.json" # for testing
```
[the remark this answers is not in the document]
Comment (BenGriffith) [reply src/crypto/transform.py:42]: ah, that shouldn't of made it into the PR. I was hardcoding blob_name with file on GCS for testing and while I removed the hardcoded value I forgot to remove the `# for testing`. I'll remove it.

File: src/crypto/transform.py [hunk 9]
```diff
@@ -1,69 +1,135 @@
...
+        )
+
+        job.result()
     
-    @time_period
+    def day_dim_rows(self) -> list:
+        rows = []
+        for key, day in enumerate(calendar.day_name, start=1):
+            row = DayDim(day_key=key, name=day)._asdict()
+            rows.append(row)
```
[the remark this answers is not in the document]
Comment (BenGriffith) [reply src/crypto/transform.py:70]: thanks for this comment. I originally started with a list comp but it was really long so I decided to break it up to be more readable.

File: src/crypto/transform.py [hunk 10]
```diff
@@ -1,69 +1,135 @@
...
+        rows = []
+        for key, month in enumerate(list(calendar.month_name)[1:], start=1):
+            row = MonthDim(month_key=key, name=month)._asdict()
+            rows.append(row)
+        return rows
     
-    def _date_dim_row(self) -> list:
-        pass
+    def date_dim_row(self) -> list:
+        d = datetime.today()
```
[the remark this answers is not in the document]
Comment (BenGriffith) [reply src/crypto/transform.py:81]: thank you

File: src/crypto/utils/constants.py (hunk 5 above)
Comment (BenGriffith) [reply src/crypto/utils/constants.py:18]: yes, it is. You're right. I removed the use of cast.

File: src/crypto/transform.py (hunk 2 above)
Comment (BenGriffith) [reply src/crypto/transform.py:96]: Good point. I see what you're saying. I'll restructure it.

File: src/crypto/transform.py (hunk 3 above)
Comment (BenGriffith) [reply src/crypto/transform.py:102]: Good call

File: src/crypto/utils/setup.py (hunk 6 above)
Comment (BenGriffith) [reply src/crypto/utils/setup.py:3]: Love this. I wasn't aware. Thank you.

File: src/crypto/transform.py (hunk 2 above)
Comment (BenGriffith) [reply src/crypto/transform.py:96]: maybe we can visit this in our next call. Was looking at it more and think I know but not completely certain.

Comment (bbelderbos) [reply src/crypto/transform.py:96]: Yep this is a good one to discuss more deeply. I will let you explain the design and we can weigh pros/cons ...

**Pull request merged**